raiden/network/rpc/client.py

```diff
@@ -820,36 +826,75 @@ def estimate_gas(
+          happen because:
+          - Another application is using the same private key.
+          - The node restarted and the `nonce` recovered by
+            `discover_next_available_nonce` was too low, which can happend
+            because:
+            - The transactions currenlty in the pool are not taken into
+              account.
+            - There was a gap in the `nonce`s of the transaction in the pool,
+              once the gap is filled a `nonce` is reused. This is most likely a
+              bug.
+        """
+
         # Exposing the JSONRPCClient with a nice name for the instance closure.
         client = self
 
-        # The class is hidden in the method to force this method to be called,
-        # this is necesary to enforce the monotonicity of the `nonce`.
         @dataclass
-        class TransactionSlotImplementation(TransactionSlot):
-            """A poor's man linear type that will check at the runtime that a nonce is
-            used only once.
+        class TransactionSlot:
+            from_address: Address
+            data: Union[SmartContractCall, ByteCode, EthTransfer]
+            eth_node: Optional[EthClient]
+            extra_log_details: Dict[str, Any]
+            startgas: int
+            gas_price: int
+            nonce: Nonce
 
-            This is necessary to avoid problems with nonce synchronization. If a nonce
-            is not used then all subsequent transactions won't be mined, or if a nonce
-            is used more than once, only one transaction will succeed while all others
-            will fail, which is currently not supported by the Raiden node.
-            """
+            def __post_init__(self) -> None:
+                self.extra_log_details.setdefault("token", str(uuid4()))
 
+                typecheck(self.from_address, T_Address)
+                typecheck(self.data, (SmartContractCall, ByteCode, EthTransfer))
+                typecheck(self.startgas, int)
+                typecheck(self.gas_price, int)
+                typecheck(self.nonce, T_Nonce)
+
+            def to_log_details(self) -> Dict[str, Any]:
+                log_details = self.data.to_log_details()
+                log_details.update(self.extra_log_details)
+                log_details.update(
+                    {
+                        "node": to_checksum_address(client.address),
+                        "from_address": to_checksum_address(self.from_address),
+                        "eth_node": self.eth_node,
+                        "startgas": self.startgas,
+                        "gas_price": self.gas_price,
+                        "nonce": self.nonce,
+                    }
+                )
+                return log_details
+
+        # The class is hidden in the method to force this method to be called,
+        # this is necessary to enforce the monotonicity of the `nonce`.
+        @dataclass
+        class TransactionSentImplementation(TransactionSent):
```

Comment:
I would call the ABC `AbstractTransactionSent` and the implementation just `AbstractTransactionSent`, but I haven't see a clear consensus for naming these things, yet.